client/agent_endpoint.go

```diff
@@ -233,3 +233,124 @@ func (a *Agent) Host(args *structs.HostDataRequest, reply *structs.HostDataRespo
 	reply.HostData = data
 	return nil
 }
+func (a *Agent) monitorExport(conn io.ReadWriteCloser) {
+	defer conn.Close()
+
+	// Decode arguments
+	var args cstructs.MonitorExportRequest
+
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&args); err != nil {
+		handleStreamResultError(err, pointer.Of(int64(500)), encoder)
+		return
+	}
+
+	// Check acl
+	if aclObj, err := a.c.ResolveToken(args.AuthToken); err != nil {
+		handleStreamResultError(err, pointer.Of(int64(403)), encoder)
+		return
+	} else if !aclObj.AllowAgentRead() {
+		handleStreamResultError(structs.ErrPermissionDenied, pointer.Of(int64(403)), encoder)
+		return
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	mon := monitor.New(512, a.c.logger, &log.LoggerOptions{})
+	if args.MockMonitor != nil {
+		mon = args.MockMonitor
+	}
```

Comment:
Even if we do decide to keep the mock monitor infrastructure in place, I'd really like if we could figure out a way to inject this at the time we create the monitor and not via request argument. Because this leaves the possibility open for a user to request we're using the mock. That's _probably_ harmless but it's not intended to be user-accessible, so I'd rather just make it impossible. (Plus security audits tend to ding us on this kind of thing.)

Reply:
So, I really really struggled with this and other than adding `if testing.Testing()` calls into the implementation code.  The only other thing I could think of would be adding something to the TestAgent config so that we could stub it out there, since the Agent is the one who creates the monitor but is also totally unaware of it. Is it fair to take that as another smell that this is the wrong place to try to exercise that code path? Or would that still be worth doing?